src/lsp4clj/server.clj

```diff
@@ -112,22 +112,81 @@
         log-title (format-error-code "Error receiving message" :internal-error)]
     (protocols.endpoint/log server :error e (str log-title "\n" message-details))))
 
-(defn ^:private receive-message
-  [server context message]
-  (let [message-type (coercer/input-message-type message)]
-    (try
+(defn thread-loop [buf-or-n f]
+  (let [ch (async/chan buf-or-n)]
+    (async/thread
       (discarding-stdout
-        (case message-type
-          (:parse-error :invalid-request)
-          (protocols.endpoint/log server :error (format-error-code "Error reading message" message-type))
-          :request
-          (protocols.endpoint/receive-request server context message)
-          (:response.result :response.error)
-          (protocols.endpoint/receive-response server message)
-          :notification
-          (protocols.endpoint/receive-notification server context message)))
-      (catch Throwable e ;; exceptions thrown by receive-response or receive-notification (receive-request catches its own exceptions)
-        (log-error-receiving server e message)))))
+        (loop []
+          (when-let [arg (async/<!! ch)]
+            (f arg)
+            (recur)))))
+    ch))
+
+(defn ^:private dispatch-input
+  "Dispatches messages received on the input-ch based on message type. Returns a
+  channel which will close after the input-ch is closed."
+  [server context input-ch]
+  (let [;; In order to process some requests and (all) notifications in series,
+        ;; the language server sometimes needs to block client-initiated input.
+        ;; If the language server sends requests during that time, it needs to
+        ;; receive responses, even though it's blocking other input. Otherwise,
+        ;; it will end up in a deadlock, where it's waiting to receive a
+        ;; response off the input-ch and the input-ch isn't being read from
+        ;; because the server is blocking input. See
+        ;; https://github.com/clojure-lsp/clojure-lsp/issues/1500.
+
+        ;; The messages all arrive in order on the input-ch so to get to the
+        ;; client's response, we have to queue whatever other messages it's
+        ;; sent. We do that by storing them in a sliding buffer. Because of the
+        ;; sliding buffer:
+        ;; * if the client sends a message which causes the language server to
+        ;;   block, and
+        ;; * if the language server sends a request during that time, and
+        ;; * if the client sends more than 100 other messages between when the
+        ;;   language server started blocking and when the client responds to
+        ;;   the language server's request,
+        ;; * then the client's earliest messages will be dropped.
```

Comment:
I think silently dropping messages could lead to confusing and hard to debug behaviour. Since both internal input channels use sliding buffers, we also effectively stop asserting back pressure on the client.

Instead of dropping messages, we could:

- Have multiple readers for `client-initiated-ch`, so that threads waiting for responses do not block the entire pipeline. This can currently be done in user-land by returning a promise from request handlers, maybe it can be generalised? Then again, this seems to just "move" the problem, as we could have all threads blocked eventually.

- Detect that buffers are full and raise an error, so that any server threads waiting for responses get unblocked. This is probably a bit tricky to implement, potentially using `alt!` and replacing the promise in `PendingRequest` with a promesa `deferred` that can be rejected for this purpose. This situation is similar to a timeout: The client did not produce a response "soon" enough.

I think I prefer the second approach, as there are no surprises (no deadlock, no silent message drops), and it seems to me that this kind of deadlock can only be prevented in handler code anyway. When a handler waits for a response, it needs to consider the situation that the response will never come.

By the way, we should also clear any `pending-sent-requests*` on shutdown anyway, see my other comment.

Reply:
Agreed, throwing an exception in that case seems better indeed

Reply:
> Have multiple readers for client-initiated-ch

As everyone prefers the second solution, we shouldn't dwell on this one. But for the record, it would cause problems, I think. If you have multiple threads reading from the `client-initiated-ch`, you could receive and start processing a refactoring via `executeCommand` before finishing a `didChange`. You'd end up refactoring an old version of the document. lsp4clj needs to enforce sequential processing sometimes, and having a single reader from the `client-initiated-ch` is an important part of that.

> Detect that buffers are full and raise an error, so that any server threads waiting for responses get unblocked. This is probably a bit tricky to implement, potentially using offer! (or alts!/alt!) and replacing the promise in PendingRequest with a promesa deferred

I agree, dropping messages isn't great. I thought about that a little as I was implementing this, but got overwhelmed considering all the possibilities. I think it'd be good to collaborate on a solution, because I'm not sure I know what to do.

I kind of see what you mean by raising exceptions, but thinking about it more carefully, we need to clarify—where should the exception be raised? In what thread(s)?

I just took a very quick pass at detecting that the buffers are full. As you suggested, we can do it by using [`offer!` instead of `>!`](https://github.com/clojure-lsp/lsp4clj/blob/ac2deb76ded6d3d0f8fced3643bfe7836152a4be/src/lsp4clj/server.clj#L183-L185) on the client- and server-initiated channels. But then what?

* We're in the [pipeline thread](https://github.com/clojure-lsp/lsp4clj/blob/ac2deb76ded6d3d0f8fced3643bfe7836152a4be/src/lsp4clj/server.clj#L176). We can stop that thread by returning instead of recurring. (Taking from an `async/thread` that has thrown an exception just returns nil, so when you detect a full buffer, throwing an exception is the same as returning nil.) From there we could do all the [cleanup](https://github.com/clojure-lsp/lsp4clj/blob/ac2deb76ded6d3d0f8fced3643bfe7836152a4be/src/lsp4clj/server.clj#L267-L293) in `start`. Then we could [deliver](https://github.com/clojure-lsp/lsp4clj/blob/ac2deb76ded6d3d0f8fced3643bfe7836152a4be/src/lsp4clj/server.clj#L294-L297) `:buffer-full` to the promise returned by `start`. But we don't have any guarantee that anyone is holding onto, or deref-ing that return value. The lsp4clj server would stop doing anything, but without cooperation from Scarlet or clojure-lsp no exception would be raised, and no error would get logged. The language server wouldn't necessarily stop running.
* Alternatively, you could log an error and do all the same stuff. 
* Alternatively, you could try to figure out how to raise exceptions elsewhere in the language server, in a way that would cause everything to wrap up. I'm not quite sure how that would work or where the exceptions would occur—perhaps another variant of the poison pill idea?

Anyway, again, I'd appreciate some help thinking through exactly what should happen.

Reply:
IMHO, it'd be nice to have a very basic solution quickly and see if this is actually a problem. Otherwise we either wait a long time to have something or we burn a bunch of time on something that barely ever happens.
So my vote would be to log and explode as much as is easily achievable.
If we see those errors actually occurring we can start spending real effort.

Reply:
I'd go with trying the following (I can also draft a PR if this helps):

- Change the `promise` in `PendingRequest` to a `promesa.core/deferred`. Now it can be delivered with `p/resolve!` or it can be put to an error state with `p/reject!`
- Write a function that rejects alls promises in the `pending-sent-request*` atom
- Change the buffer type to a normal fixed-size buffer
- Replace the `>!` in the pipeline with:

  ```clojure
  (when-not (async/offer! x-initiated-ch the-msg)
    ;; Buffers full. Fail any waiting pending requests and block.
    (fail-pending-sent-requests (ex-info "response not received in time" {}))
    ;; Try again, but block this time, to apply back pressure to our client
    (async/>! x-initiated-ch the-msg))
  ```

This will trigger the exception in any handler threads that are waiting for a response that we could not deliver. In handler code, this would be very similar to a timeout. I would not fail the pipeline, but as shown fall back to blocking, so that we can apply back pressure.

The function `fail-pending-sent-requests` could be reused in the clean-up phase, to interrupt any waiting threads.

Reply:
A nice side-effect of using a promesa promise inside `PendingRequest`: We could change it to be a promise itself.

We can piggy-back on the underlying `CompletableFuture`, replacing the `cancelled?` atom with `p/cancel!` and `p/cancelled?`, and implementing `IPromiseFactory` ourselves, returning the underlying promise.

Now users can use promesa to chain callbacks to the request, eliminating the need for blocking in some situations.